[quant][pt2e] Introduce QuantizationAnnotation API

jerryzh168 opened the pull request:
Summary:
This diff adds QuantizationAnnotation and also refactors the existing annotation to use this object

```
dataclass
class QuantizationAnnotation:
  # How some input nodes should be quantized, expressed as QuantizationSpec
  # a map from torch.fx.Node to QuantizationSpec
  input_qspec_map: Dict[Node, QuantizationSpec]

  # How the output of this node is quantized, expressed as QuantizationSPec
  output_qspec: QuantizationSpec

class QuantizationSpec:
    dtype: torch.dtype
    is_dynamic: bool = False
    quant_min: Optional[int] = None
    quant_max: Optional[int] = None
    qscheme: Optional[torch.qscheme] = None
    ch_axis: Optional[int] = None
    # TODO: follow up PR will add this
    # Kind of observer such as MinMaxObserver, PerChannelHistogramObserver etc.
    # observer_or_fake_quant_type: Union[ObserverBase, FakeQuantizeBase]
```

Example after full refactor:

```
int8_qspec = QuantizationSpec(dtype=torch.int8, ...)
weight_qspec = QuantizationSpec(dtype=torch.int8, ...)
conv_node["quantization_annotation"] = QuantizationAnnotation(
    input_qspec_map={input_node: int8_qspec, weight_node: weight_qspec}
    output_qspec=int8_qspec,
)
```

Note: right now input_qspec_map and output_qspec map are still using observer and fake quant constructors.
Follow up PR: change the input_qspec_map and output_qspec to use QuantizationSpec directly

Test Plan:
```
buck2 test mode/optcaffe2/test:quantization_pt2e -- --exact 'caffe2/test:quantization_pt2e - test_resnet18_with_quantizer_api (quantization.pt2e.test_quantize_pt2e.TestQuantizePT2EModels)'
```

Differential Revision: D45895027

leslie-fang-intel commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -445,18 +452,24 @@ def _annotate_conv2d(
+        weight = conv_node.args[1]
+        assert isinstance(weight, Node)
+        bias = conv_node.args[2]
+        assert isinstance(bias, Node)
```
Maybe we can consider to add a function such as `_is_all_fx_node(node_list)`? It can reduce the lines of duplicated code.

jerryzh168 replied:
this annotation code will be changed in the future to use module_partition API, and this may not be needed

leslie-fang-intel replied:
Got it. Thanks.

leslie-fang-intel commented on torch/ao/quantization/_pt2e/quantizer/quantizer.py:
```diff
@@ -98,6 +103,29 @@ class QuantizationConfig:
 )
 
+@dataclass
+class QuantizationAnnotation:
```
If we want to add more fields only for customized backend recipe. Take IPEX recipes for example, should we consider a `IPEXQuantizationAnnotation` class which inherit this class with added fields?

jerryzh168 replied:
I think it's better for you to define a class separately and attach it to meta field of node, e.g.:

```
node.meta["ipex_annotation"] = IPEXAnnotation(...)
```

leslie-fang-intel replied:
But looks like it's hardcoded checkout of filed `quantization_annotation` in node.meta  https://github.com/pytorch/pytorch/blob/433672ec9cc4fe11bfb1e53b35601b76316899d4/torch/ao/quantization/fx/prepare.py#L527

jerryzh168 replied:
it will be preserved and should not interact with prepare

jerryzh168 committed:
[quant][pt2e] Introduce QuantizationAnnotation API (#101708)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/101708

This diff adds QuantizationAnnotation and also refactors the existing annotation to use this object

```
dataclass
class QuantizationAnnotation:
  # How some input nodes should be quantized, expressed as QuantizationSpec
  # a map from torch.fx.Node to QuantizationSpec
  input_qspec_map: Dict[Node, QuantizationSpec]

  # How the output of this node is quantized, expressed as QuantizationSPec
  output_qspec: QuantizationSpec

class QuantizationSpec:
    dtype: torch.dtype
    is_dynamic: bool = False
    quant_min: Optional[int] = None
    quant_max: Optional[int] = None
    qscheme: Optional[torch.qscheme] = None
    ch_axis: Optional[int] = None
    # TODO: follow up PR will add this
    # Kind of observer such as MinMaxObserver, PerChannelHistogramObserver etc.
    # observer_or_fake_quant_type: Union[ObserverBase, FakeQuantizeBase]
```

Example after full refactor:

```
int8_qspec = QuantizationSpec(dtype=torch.int8, ...)
weight_qspec = QuantizationSpec(dtype=torch.int8, ...)
conv_node["quantization_annotation"] = QuantizationAnnotation(
    input_qspec_map={input_node: int8_qspec, weight_node: weight_qspec}
    output_qspec=int8_qspec,
)
```

Note: right now input_qspec_map and output_qspec map are still using observer and fake quant constructors.
Follow up PR: change the input_qspec_map and output_qspec to use QuantizationSpec directly

Test Plan:
```
buck2 test mode/opt caffe2/test:quantization_pt2e -- --exact 'caffe2/test:quantization_pt2e - test_resnet18_with_quantizer_api (quantization.pt2e.test_quantize_pt2e.TestQuantizePT2EModels)'
```

Reviewed By: kimishpatel

Differential Revision: D45895027

fbshipit-source-id: 6ac2ff1b150171476ebcca14850bc134cdbbbd8e
